src/webhooks/events/invoices/InvoiceCompletedHandler.ts

```
@@ -228,4 +228,97 @@ export class InvoiceCompletedHandler {
+      const matchingTier = userTiers.find((userTier) => {
+        const hasWorkspaces = userTier.featuresPerService[Service.Drive].workspaces.enabled;
+        return isBusinessPlan ? hasWorkspaces : !hasWorkspaces;
+      });
+      if (matchingTier) {
+        await this.tiersService.updateTierToUser(userId, matchingTier.id, tierId);
+      } else {
+        await this.tiersService.insertTierToUser(userId, tierId);
+      }
+    } catch (error) {
+      this.logger.error(`Error while updating or inserting the user-tier relationship. Error: ${error}`);
+    }
+  }
+
+  /**
+   * Handle the relationship between the user and the coupon.
+   * If the invoice has a discount that we track internally and the coupon is not the free trial one,
+   * store the coupon id in the user's `usedCoupons` field.
+   * @param userUuid The uuid of the user
+   * @param invoice The invoice where the discount is applied
+   * @param invoiceLineItem The invoice line item which has the discount
+   * @param isLifetimePlan Whether the invoice is for a lifetime plan
+   */
+  private async handleUserCouponRelationship({
+    userUuid,
+    invoice,
+    invoiceLineItem,
+    isLifetimePlan,
+  }: {
+    userUuid: string;
+    invoice: Stripe.Invoice;
+    invoiceLineItem: Stripe.InvoiceLineItem;
+    isLifetimePlan: boolean;
+  }): Promise<void> {
+    try {
+      const userData = await this.usersService.findUserByUuid(userUuid);
+
+      const areDiscounts = isLifetimePlan ? invoiceLineItem.discounts.length > 0 : !!invoice.discount?.coupon;
+      if (areDiscounts) {
+        const coupon = isLifetimePlan
+          ? (invoiceLineItem.discounts[0] as Stripe.Discount).coupon
+          : invoice.discount?.coupon;
+
+        if (coupon) {
+          await this.usersService.storeCouponUsedByUser(userData, coupon.id);
+        }
+      }
+    } catch (err) {
+      const error = err as Error;
+      if (!(err instanceof CouponNotBeingTrackedError)) {
```

else the error is lost